### ISSUE TITLE
Incorrect implementation of YEARFRAC

### ISSUE THREAD

**Issue by michaelhays** [dgorissen/pycel#111]
The implementation of YEARFRAC for 30/360 US does not match the [spec](https://cbonds.com/glossary/30-360-us/).

[This line](https://github.com/dgorissen/pycel/blob/ab80150780358e4b95554fa4df3a6763618395a4/src/pycel/lib/date_time.py#L409):

```py
d2 = max(d2, 30) if d1 == 30 else d2
```

I believe should use `min` instead of `max`:

```py
d2 = min(d2, 30) if d1 == 30 else d2
```

This results in vastly different results for some dates, like 2019-12-31 (43830) to 2020-01-01 (43831):

```py
from pycel.lib.date_time import yearfrac

# Correct value
yearfrac(43830, 43831, 0) == 1/360

# Actual value
yearfrac(43830, 43831, 0) == 30/360
```

**Comment by dgorissen** [comment]
Thanks Micheal. Please issue a PR (incl tests) if you can.

**Comment by michaelhays** [comment]
Will do, may not have time this week but I'll get to it by the weekend.

**Comment by michaelhays** [comment]
## TL;DR

Excel's YEARFRAC claims to use the "US (NASD)" day count convention, but it appears to use an incorrect implementation. Google Sheets uses a different (seemingly correct) implementation, and thus calculates different values for certain edge cases.

So the main question is: should the code in this library match the Excel implementation exactly, or the actual "US (NASD)" implementation?

What do you think @dgorissen? I'm thinking it's probably best to match Excel, even if it's occasionally wrong. There's some pseudocode at the bottom of this comment which I believe matches Excel's version. If you agree, I'll change the pycel implementation to that logic, with some tests to verify it matches Excel.

## Details

The [Excel YEARFRAC documentation](https://support.microsoft.com/en-us/office/yearfrac-function-3844141e-c76d-4143-82b6-208454ddc6a8) claims to use the "US (NASD)" version of 30/360. I'm not sure where to find the source definition of this convention, but a found a few places ([1](https://help.anaplan.com/en/cb6a2f39-8109-416e-97d7-7614a19965da-Day-count-conventions), [2](https://www.fundmanagersoftware.com/help/def_accrued_interest.html)) that define it as:

> Rule 1: If the investment is EOM and (Date1 is the last day of February) and (Date2 is the last day of February), then change D2 to 30.
Rule 2: If the investment is EOM and (Date1 is the last day of February), then change D1 to 30.
Rule 3: If D2 is 31 and D1 is 30 or 31, then change D2 to 30.
Rule 4: If D1 is 31, then change D1 to 30.

Wikipedia simply lists this as [30/360 US](https://en.wikipedia.org/wiki/Day_count_convention#30/360_US). From the [2006 ISDA Definitions](https://web.archive.org/web/20150924031602/http://www.hsbcnet.com/gbm/attachments/standalone/2006-isda-definitions.pdf) for 30/360 (Sections 4.16 f-h), this is closest to 30E/360 ISDA (Section 4.16 h), but not identical.

### Examples

[Here](https://www.isda.org/2008/12/22/30-360-day-count-conventions/) is a helpful document of a few different 30/360 day count conventions, with example calculations for tricky edge cases, which we can use for testing.

YEARFRAC seems to get most calculations correct, except intervals with the start date as the last day in February (Note, calculations were performed in Excel online):

Start Date | End Date | YEARFRAC | Days
-- | -- | -- | --
2/28/2007 | 3/31/2007 | 0.086111111 | 31
2/28/2007 | 8/31/2007 | 0.502777778 | 181
2/29/2008 | 3/31/2008 | 0.086111111 | 31
2/29/2008 | 8/31/2008 | 0.502777778 | 181

Here is that same table in Google Sheets:

Start Date | End Date | YEARFRAC | Days
-- | -- | -- | --
2/28/2007 | 3/31/2007 | 0.083333333 | 30
2/28/2007 | 8/31/2007 | 0.5 | 180
2/29/2008 | 3/31/2008 | 0.083333333 | 30
2/29/2008 | 8/31/2008 | 0.5 | 180

The Google Sheets calculations match the set of rules described above, but the Excel online calculations don't.

### dwheeler document

[This 2008 article](http://web.archive.org/web/20200914232323/https://dwheeler.com/yearfrac/) highlights inaccuracies with Excel's YEARFRAC.

It includes some pseudocode for each Basis of the actual implementation of Excel's YEARFRAC. Here is the pseudocode for Basis 0:

```
def basis0(date1,date2):
  # Swap so date1 <= date2 in all cases:
  if date1 > date2:
    date1, date2 = date2, date1
  if date1 == date2:
    return 0.0
  # Change day-of-month for purposes of calculation.
  date1day, date1month, date1year = date1.day, date1.month, date1.year
  date2day, date2month, date2year = date2.day, date2.month, date2.year
  if (date1day == 31 and date2day == 31):
    date1day = 30
    date2day = 30
  elif date1day == 31: date1day = 30
  elif (date1day == 30 and date2day == 31): date2day = 30
  # Note: If date2day==31, it STAYS 31 if date1day < 30.
  # Special fixes for February:
  elif (date1month == 2 and date2month == 2 and last_day_of_month(date1)
      and last_day_of_month(date2)):
     date1day = 30  # Set the day values to be equal
     date2day = 30
  elif date1month == 2 and last_day_of_month(date1):
     date1day = 30 # "Illegal" Feb 30 date.
  daydiff360 = ( (date2day + date2month * 30 + date2year * 360) -
                 (date1day + date1month * 30 + date1year * 360))
  return daydiff360 / 360
```

Indeed, this does produce the correct results for the Excel table above. Particularly, for every row, it sets D1 = 30, but does not change D2.

If this pseudocode is accurate, it seems likely that Excel simply uses an incorrect implementation of 30/360 (or, you could say, accidentally invented its own).
